test/integration/src/test/java/io/pravega/test/integration/BackUpRecoveryTest.java

```diff
@@ -0,0 +1,1220 @@
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.retry.ExponentialBackoffRetry;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.File;
+import java.net.URI;
+import java.nio.file.Files;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import static io.pravega.common.concurrent.Futures.loop;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Integration test to verify data recovery.
+ * Recovery scenario: when data written to Pravega is already flushed to the long term storage.
+ * Tests replicate different environments for data recovery.
+ */
+@Slf4j
+public class BackUpRecoveryTest extends ThreadPooledTestSuite {
+    private static final Duration TIMEOUT = Duration.ofMillis(100 * 10000);
+    private static final Duration READ_TIMEOUT = Duration.ofMillis(1000);
+    private static final Duration TRANSACTION_TIMEOUT = Duration.ofMillis(10000);
+
+    /**
+     * Write {@link #TOTAL_NUM_EVENTS} events to verify recovery.
+     */
+    private static final int TOTAL_NUM_EVENTS = 300;
+
+    private static final Random RANDOM = new Random(1234);
+
+    /**
+     * Scope and streams to read and write events.
+     */
+    private static final String SCOPE = "testRecoveryScope";
+    private static final String STREAM1 = "testRecoveryStream" + RANDOM.nextInt(Integer.MAX_VALUE);
+    private static final String EVENT = "12345";
+    private final ScalingPolicy scalingPolicy = ScalingPolicy.fixed(1);
+    private final StreamConfiguration config = StreamConfiguration.builder().scalingPolicy(scalingPolicy).build();
+
+    private final AtomicLong timer = new AtomicLong();
+
+    private FileSystemSimpleStorageFactory storageFactory;
+
+    private BookKeeperLogFactory dataLogFactory;
+
+    @After
+    public void tearDown() throws Exception {
+        if (this.dataLogFactory != null) {
+            this.dataLogFactory.close();
+        }
+        timer.set(0);
+    }
+
+    @Override
+    protected int getThreadPoolSize() {
+        return 100;
+    }
+
+    /**
+     * Sets up a new BookKeeper & ZooKeeper.
+     */
+    public static class BookKeeperRunner implements AutoCloseable {
```

Comment:
why not re-use BookkeeperRunner, PravegaRunner, SegmentSTore runner from LocalServicestarter?

Reply:
updated to reuse LocalServicestarter